### PR TITLE
Switch the baseurl with `set_api`

### LINKED ISSUE #24: Add a way to change the BASEURL
There are several (I am unclear how many) mirrors for Helioviewer and people might want to use a mirror instead of the main URL. So we will need to add a way to change this either as a keyword, a configuration option or even a system environment variable.

### PR DESCRIPTION
**Pull request by akash5100**
* Closes #24

### CODE REVIEW

File: hvpy/api_groups/jpeg2000/tests/test_get_jp2_header.py [hunk 1]
```diff
@@ -23,3 +26,17 @@ def test_error_handling():
...
+def test_set_api():
+    params = getJP2HeaderInputParameters(id=9838343)
+    assert params.url == "https://api.helioviewer.org/v2/getJP2Header/"
+
+    hvpy.set_api("https://api.beta.helioviewer.org/")
+    params = getJP2HeaderInputParameters(id=9838343)
+    assert params.url == "https://api.beta.helioviewer.org/getJP2Header/"
+
+    os.environ["URL"] = "https://localhost:3000/"
+    hvpy.set_api()
```
Comment (nabobalis) [inline hvpy/api_groups/jpeg2000/tests/test_get_jp2_header.py:40]: So I have to do this to change the URL?

File: hvpy/io.py [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
 from enum import Enum, auto
-from typing import Any, Dict
+from typing import Any, Dict, Optional
 
 from pydantic import BaseModel
 
-__all__ = ["HvpyParameters", "OutputType"]
+__all__ = ["HvpyParameters", "OutputType", "set_api"]
 
-BASE_URL = "https://api.helioviewer.org/v2/"
```
Comment (nabobalis) [inline hvpy/io.py:8]: On the surface, I don't see why you can't just set the value here if the OS environmental variable exists?

File: hvpy/api_groups/jpeg2000/tests/test_get_jp2_header.py (hunk 1 above)
Comment (akash5100) [reply hvpy/api_groups/jpeg2000/tests/test_get_jp2_header.py:40]: We decided this in the meeting on, 14th [here are the notes](https://github.com/Helioviewer-Project/python-api/wiki/Meeting-Notes#14-july-830-ist) and (pseudo code) by Daniel.

Comment (nabobalis) [reply hvpy/api_groups/jpeg2000/tests/test_get_jp2_header.py:40]: I see, thanks for the context. 

I still don't understand why a person should need to run `hvpy.set_api()` if they have set a OS environment.

File: hvpy/io.py [hunk 3]
```diff
@@ -54,4 +55,17 @@ def url(self) -> str:
...
         """
-        return BASE_URL + self.__class__.__name__[:-15] + "/"
+        return base_url + self.__class__.__name__[:-15] + "/"
+
+
+def set_api(url: Optional[str] = None) -> None:
+    """
+    Sets the base URL for the Helioviewer python API.
+    """
+    global base_url
```
Comment (nabobalis) [inline hvpy/io.py:65]: We need a different way that isn't a global variable.

File: hvpy/api_groups/jpeg2000/tests/test_get_jp2_header.py (hunk 1 above)
Comment (akash5100) [reply hvpy/api_groups/jpeg2000/tests/test_get_jp2_header.py:40]: Yes, I know. I tried to do what you said but I was not able to unless I ran the `set_api` function.

File: hvpy/io.py (hunk 3 above)
Comment (akash5100) [reply hvpy/io.py:65]: Right.

File: hvpy/io.py [hunk 4]
```diff
@@ -1,11 +1,12 @@
...
-from typing import Any, Dict
+from typing import Any, Dict, Optional
 
 from pydantic import BaseModel
 
-__all__ = ["HvpyParameters", "OutputType"]
+__all__ = ["HvpyParameters", "OutputType", "set_api"]
 
-BASE_URL = "https://api.helioviewer.org/v2/"
+base_url = "https://api.helioviewer.org/v2/"
```
Comment (akash5100) [inline hvpy/io.py:9]:
```suggestion
if "URL" in os.environ:
    base_url = os.environ.get("URL")
else:
    base_url = "https://api.helioviewer.org/v2/"
```

So will it be like this?

Comment (nabobalis) [reply hvpy/io.py:9]: We can set the second argument to `get` to be the main URL and we won't need the if statement.

It should also be renamed to something other than `URL`, its too vague.

File: hvpy/io.py [hunk 5]
```diff
@@ -54,4 +53,11 @@ def url(self) -> str:
...
         """
-        return BASE_URL + self.__class__.__name__[:-15] + "/"
+        base_url = os.environ.get("HVPY_BASE_URL", default="https://api.helioviewer.org/v2/")
+        return base_url + self.__class__.__name__[:-15] + "/"
+
+
+def set_api(url: Optional[str] = None) -> None:
+    """
+    Sets the base URL for the Helioviewer python API.
+    """
```
[the remark this answers is not in the document]
Comment (nabobalis) [reply hvpy/io.py:63]: On the surface, I would still have the `BASE_URL` at the top of the file like before, have its v value set from either the OS environment or set to the default. 

You have `set_api` set a private _url attribute on the class and then you check if it exists, if not, use the BASE_URL.

[the remark this answers is not in the document]
Comment (akash5100) [reply hvpy/io.py:63]: If I put this line at the top:
```py
BASE_URL = os.environ.get("HVPY_BASE_URL", default="https://api.helioviewer.org/v2/")
```
the assertion fails. 
But if this line is put inside a function, for example:
```py
def get_url() -> str:
    BASE_URL = os.environ.get("HVPY_BASE_URL", default="https://api.helioviewer.org/v2/")
    return BASE_URL
```
the assertion passes, weird. 

```sh
        os.environ["HVPY_BASE_URL"] = "https://localhost:3000/"
        params = getJP2HeaderInputParameters(id=9838343)
>       assert params.url == "https://localhost:3000/getJP2Header/"
E       AssertionError: assert 'https://api....getJP2Header/' == 'https://loca...getJP2Header/'
E         - https://localhost:3000/getJP2Header/
E         + https://api.helioviewer.org/v2/getJP2Header/

hvpy/api_groups/jpeg2000/tests/test_get_jp2_header.py:36: AssertionError
```

[the remark this answers is not in the document]
Comment (nabobalis) [reply hvpy/io.py:63]: That is expected.

We might want a get_api_url and a set_api_url function.

[the remark this answers is not in the document]
Comment (akash5100) [reply hvpy/io.py:63]: Done, Thanks.

File: hvpy/io.py [hunk 6]
```diff
@@ -54,4 +53,19 @@ def url(self) -> str:
...
+    Returns the base URL for all API calls.
+    """
+    return os.environ.get("HVPY_BASE_URL", default="https://api.helioviewer.org/v2/")
+
+
+def set_api_url(private_url: str) -> None:
+    """
+    Sets the base URL for all API calls.
+    """
+    os.environ["HVPY_BASE_URL"] = private_url
```
Comment (nabobalis) [inline hvpy/io.py:71]: I am not sure we want to override this.

File: hvpy/io.py [hunk 7]
```diff
@@ -54,4 +53,19 @@ def url(self) -> str:
         """
         Final API endpoint URL.
         """
-        return BASE_URL + self.__class__.__name__[:-15] + "/"
+        base_url = get_api_url()
+        return base_url + self.__class__.__name__[:-15] + "/"
```
Comment (akash5100) [inline hvpy/io.py:57]:
```suggestion
        os.environ.clear()
        return base_url + self.__class__.__name__[:-15] + "/"
```
Tested this locally, and it passes all the tests. 
Should we do it?

Comment (nabobalis) [reply hvpy/io.py:57]: I don't think we should modifying a user's environment when someone uses hvpy.

File: hvpy/io.py [hunk 8]
```diff
@@ -54,4 +53,23 @@ def url(self) -> str:
...
+
+
+def get_api_url() -> str:
+    """
+    Returns the base URL for all API calls.
+    """
+    if "PRIVATE_URL" in os.environ:
+        base_url = os.environ["PRIVATE_URL"]
+        del os.environ["PRIVATE_URL"]
+        return base_url
```
Comment (nabobalis) [inline hvpy/io.py:67]: I still do not like this.

Comment (akash5100) [reply hvpy/io.py:67]: I am sorry for this. But I don't know, how do we use `set_api_url()` to modify the base URL of the `HvpyParameters` class?

Comment (nabobalis) [reply hvpy/io.py:67]: Please don't apologize. I don't have a good idea at the moment either.

Comment (dgarciabriseno) [reply hvpy/io.py:67]: Let's make base_url a file level variable `_base_url`

Define it to either `api.helioviewer.org` or whatever is in the environment.

Something like
```python
_base_url = os.environ['PRIVATE_URL'] if 'PRIVATE_URL' in os.environ else "https://api.helioviewer.org/v2/"
# or if you don't like python's strange ternary
_base_url = "https://api.helioviewer.org/v2/"
if 'PRIVATE_URL' in os.environ:
  _base_url = os.environ['PRIVATE_URL']
```

This meets the first requirement of starting with our default or the user's environment.

`set_api_url` modified the file level `_base_url`

`HvpyParameters` uses the file level `_base_url`

That should do it.

File: hvpy/io.py [hunk 9]
```diff
@@ -54,4 +56,14 @@ def url(self) -> str:
         """
         Final API endpoint URL.
         """
-        return BASE_URL + self.__class__.__name__[:-15] + "/"
+        if "PRIVATE_URL" in os.environ:
+            return os.environ["PRIVATE_URL"] + self.__class__.__name__[:-15] + "/"
```
[the remark this answers is not in the document]
Comment (dgarciabriseno) [reply hvpy/io.py:60]: This needs to move to initialize `_base_url`

File: hvpy/io.py [hunk 10]
```diff
@@ -54,4 +56,24 @@ def url(self) -> str:
...
+        return get_api_url() + self.__class__.__name__[:-15] + "/"
+
+
+def get_api_url() -> str:
+    """
+    Returns the base API URL.
+    """
+    if _base_url:
+        return _base_url
+    if "PRIVATE_URL" in os.environ:
```
Comment (nabobalis) [inline hvpy/io.py:69]: This variable name is still too vague.

Comment (dgarciabriseno) [reply hvpy/io.py:69]: Yes, need to change PRIVATE_URL to something more helioviewer specific.

**Review by dgarciabriseno — CHANGES_REQUESTED** [review]
Change `PRIVATE_URL` to something more specific to hvpy. Remember the environment contains variables that are passed to every program that runs on a system, so "PRIVATE_URL" is too vague.

Let's use `HELIOVIEWER_API_URL`

**Review by dgarciabriseno — APPROVED** [review]

**Review by nabobalis — APPROVED** [review]

**Pull request merged**